File: packages/offline_stt/darwin/Classes/OfflineSttApiImpl.swift
// OfflineSttApiImpl.swift
// `OfflineSttHostApi`(Pigeon生成、Pigeon.g.swift)の実装本体。
// design.md §4.2、requirements.md FR-1〜FR-3 FR-6(Issue #33〜#39)に対応する。
//
// ## 同期プロトコルと非同期Speech APIの橋渡しについて
// `pigeons/offline_stt_events.dart` の `checkModel` には `@async` を
// 付与しているため、Pigeonが生成するSwiftプロトコルは
// `completion: @escaping (Result<ModelState, Error>) -> Void` を受け取る
// 非同期シグネチャである。これにより `AssetInventory` / `SpeechTranscriber`
// のSwift ConcurrencyのasyncAPIを、Flutterのプラットフォームスレッドを
// ブロックせずに`Task`経由でそのまま呼び出せる(以前は`SyncBridge.swift`の
// `runBlocking`でスレッドをブロックしていたが、ANR・デッドロックの危険が
// あるため廃止した)。
// - `downloadModel` / `transcribeFile` / `cancel` は「開始のみ」を表す
//   契約(`pigeons/offline_stt_events.dart`の各メソッドのドキュメント
//   コメント参照)であり `@async` を付与していないため、引き続き同期
//   シグネチャ(`throws`)である。`Task`を起動して即座に返り、結果は
//   `segments` / `downloadProgress` のEventChannelで非同期に配信する。
#if os(macOS)
  import FlutterMacOS
#else
  import Flutter
#endif
import Foundation
import Speech

final class OfflineSttApiImpl: NSObject, OfflineSttHostApi {
  private let segmentsWrapper: SegmentsEventWrapper
  private let downloadProgressWrapper: DownloadProgressEventWrapper

  private var currentTranscriptionTask: Task<Void, Never>?
  private var currentDownloadTask: Task<Void, Never>?

  init(segmentsWrapper: SegmentsEventWrapper, downloadProgressWrapper: DownloadProgressEventWrapper) {
    self.segmentsWrapper = segmentsWrapper
    self.downloadProgressWrapper = downloadProgressWrapper
  }

  // MARK: - OfflineSttHostApi (#34 #35)

  func checkModel(locale: String, completion: @escaping (Result<ModelState, Error>) -> Void) {
    // #34 OSバージョンゲート: 26未満では`unavailable`を返す。`if #available`
    // による分岐のため、パッケージ全体は26未満でもコンパイルが通る
    // (design.md §4.2)。
    guard #available(macOS 26.0, iOS 26.0, *) else {
      completion(.success(.unavailable))
      return
    }
    Task {
      let state = await ModelAvailability.checkModel(localeIdentifier: locale)
      completion(.success(state))
    }
  }

  func supportedLocales(completion: @escaping (Result<[String], Error>) -> Void) {
    // OSバージョンゲート。`checkModel` は26未満で `unavailable`(FR-1の
    // 終端状態)を返せるが、一覧には「対応ロケールが無い」を表す正しい値が
    // 無いため、ここは明示的なエラーにする(空配列は「このOSは1言語も
    // 扱えない」と誤解されるため返さない)。
    guard #available(macOS 26.0, iOS 26.0, *) else {
      completion(
        .failure(
          DarwinTranscribeError.deviceUnsupported(
            "対応ロケールの列挙には macOS 26 / iOS 26 以上が必要である。"
          ).asPigeonError))
      return
    }
    Task {
      do {
        completion(.success(try await ModelAvailability.supportedLocales()))
      } catch let error as DarwinTranscribeError {
        completion(.failure(error.asPigeonError))
      } catch {
        completion(.failure(DarwinTranscribeError.platformError("\(error)").asPigeonError))
      }
    }
  }

  func downloadModel(locale: String) throws {
    guard #available(macOS 26.0, iOS 26.0, *) else {
      // design.md §3細則3: downloadable以外(ここではOS非対応=unavailable)
      // で呼ばれた場合は、状態を変化させず何もemitせず完了する。
      downloadProgressWrapper.sendEndOfStream()
      return
    }
    currentDownloadTask?.cancel()
    currentDownloadTask = Task { [weak self] in
      await self?.runDownload(locale: locale)
    }
  }

  func transcribeFile(request: TranscribeRequest) throws {
    guard #available(macOS 26.0, iOS 26.0, *) else {
      segmentsWrapper.sendError(.deviceUnsupported("OS 26未満"))
      return
    }
    // design.md §3の同時セッション排他(v1では1本まで)はDart側
    // `TranscribeSessionGuard`が担うため、通常この時点で前のTaskが
    // 生きていることはない。念のため多重起動を防ぐ。
    currentTranscriptionTask?.cancel()
    currentTranscriptionTask = Task { [weak self] in
      await self?.runTranscription(request: request)
    }
  }

  func cancel() throws {
    currentTranscriptionTask?.cancel()
    currentDownloadTask?.cancel()
  }

  /// `segments` EventChannel自体がキャンセルされた場合の保険
  /// (EventChannelWrappers.swift のドキュメントコメント参照)。
  ///
  /// 停止するのは文字起こしTaskのみである。2本のEventChannelは独立して
  /// いるため、片方の購読解除でもう片方を巻き添えにしてはならない
  /// (例: ja-JPの文字起こし中にen-USのダウンロードStreamの購読を
  /// 解除しても、文字起こしは継続しなければならない)。明示的な
  /// `cancel()` HostApiは従来どおり両方を停止する。
  func cancelTranscriptionFromEventChannel() {
    currentTranscriptionTask?.cancel()
  }

  /// `downloadProgress` EventChannel自体がキャンセルされた場合の保険。
  /// 停止するのはダウンロードTaskのみである(理由は
  /// `cancelTranscriptionFromEventChannel()` のコメント参照)。
  func cancelDownloadFromEventChannel() {
    currentDownloadTask?.cancel()
  }

  // MARK: - ダウンロード(#35)

  @available(macOS 26.0, iOS 26.0, *)
  private func runDownload(locale: String) async {
    // design.md §3細則3: downloadable以外で呼ばれた場合は状態を変化させず
    // 何もemitせず完了する。
    let state = await ModelAvailability.checkModel(localeIdentifier: locale)
    guard state == .downloadable else {
      downloadProgressWrapper.sendEndOfStream()
      return
    }

    downloadProgressWrapper.send(fraction: 0.0, completed: false)
    do {
      try await ModelAcquisition.run(localeIdentifier: locale) { [weak self] fraction in
        self?.downloadProgressWrapper.send(fraction: fraction, completed: false)
      }
      downloadProgressWrapper.send(fraction: 1.0, completed: true)
      downloadProgressWrapper.sendEndOfStream()
    } catch let error as DarwinTranscribeError {
      downloadProgressWrapper.sendError(error)
      downloadProgressWrapper.sendEndOfStream()
    } catch is CancellationError {
      downloadProgressWrapper.sendError(.cancelled)
      downloadProgressWrapper.sendEndOfStream()
    } catch {
      downloadProgressWrapper.sendError(.platformError("\(error)"))
      downloadProgressWrapper.sendEndOfStream()
    }
  }

  // MARK: - 文字起こし(#36 #37 #38)

  @available(macOS 26.0, iOS 26.0, *)
  private func runTranscription(request: TranscribeRequest) async {
    // design.md §2.2: `playbackRate`はWeb専用オプションであり、Darwinでは
    // 無視する。`pigeons/offline_stt_events.dart`のTranscribeRequestに
    // そもそも`playbackRate`フィールドが存在しない(現行ブランチのPigeon
    // スキーマにはPR#74で追加予定のplaybackRateが未反映であるため)ため、
    // 本メソッドの引数`request`は元々`path`/`locale`のみで構成されており、
    // 無視は自動的に満たされる。
    let session = TranscriptionSession(
      localeIdentifier: request.locale,
      filePath: request.path,
      onSegment: { [weak self] segment in
        self?.segmentsWrapper.send(segment)
      }
    )
    do {
      try await session.run()
      segmentsWrapper.sendEndOfStream()
    } catch let error as DarwinTranscribeError {
      segmentsWrapper.sendError(error)
    } catch is CancellationError {
      segmentsWrapper.sendError(.cancelled)
    } catch {
      segmentsWrapper.sendError(.platformError("\(error)"))
    }
  }
}

File: packages/offline_stt/android/src/main/kotlin/com/moongift/offline_stt/OfflineSttApiImpl.kt
// OfflineSttApiImpl.kt
// `OfflineSttHostApi`(Pigeon生成、Pigeon.g.kt)の実装本体。
// design.md §4.3(wt73版)、requirements.md FR-1〜FR-3 FR-6(Issue #43〜#48)
// に対応する(Darwin実装のOfflineSttApiImpl.swiftのKotlin版)。
//
// `checkModel`/`downloadModel`/`transcribeFile`/`cancel`いずれも
// `SpeechRecognizer.createOnDeviceSpeechRecognizer()`を要求するため、
// design.md §4.3(wt73版)注記9のとおりメインスレッドから呼び出す必要がある。
// `OfflineSttHostApi`のメソッドはFlutterのプラットフォームスレッド
// (Android上はメインスレッドと同一)から呼ばれるため、追加のスレッド切替は
// 不要である。
package com.moongift.offline_stt

import android.content.Context
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

class OfflineSttApiImpl(
    private val context: Context,
    private val segmentsWrapper: SegmentsEventWrapper,
    private val downloadProgressWrapper: DownloadProgressEventWrapper,
) : OfflineSttHostApi {

    // Dispatchers.Main.immediate: design.md §4.3(wt73版)注記9のとおり
    // SpeechRecognizerはメインスレッドから操作する必要があるため。
    // SupervisorJob: checkModel呼び出しがdownloadModel/transcribeFileの
    // 失敗で巻き添えにならないようにする。
    private val mainScope = CoroutineScope(Dispatchers.Main.immediate + SupervisorJob())

    private var currentTranscriptionJob: Job? = null
    private var currentDownloadJob: Job? = null

    override fun checkModel(locale: String, callback: (Result<ModelState>) -> Unit) {
        mainScope.launch {
            // **例外を必ず callback へ渡す。** `mainScope.launch` の中で
            // 例外が外へ抜けると、Pigeon のコールバックが呼ばれないまま
            // コルーチンが異常終了し、Dart 側は応答を待ち続けるか
            // アプリごと落ちる(実機で確認した)。
            try {
                callback(Result.success(ModelAvailability.checkModel(context, locale)))
            } catch (e: CancellationException) {
                throw e
            } catch (e: AndroidTranscribeError) {
                callback(Result.failure(e.toFlutterError()))
            } catch (e: Exception) {
                callback(
                    Result.failure(
                        AndroidTranscribeError.PlatformError("$e").toFlutterError(),
                    ),
                )
            }
        }
    }

    override fun supportedLocales(callback: (Result<List<String>>) -> Unit) {
        mainScope.launch {
            // `checkModel` と同じ理由で例外を必ず callback へ渡す。
            // `mainScope.launch` の中で例外が外へ抜けると、Pigeon の
            // コールバックが呼ばれないままコルーチンが異常終了し、Dart 側は
            // 応答を待ち続けるかアプリごと落ちる。
            try {
                callback(Result.success(ModelAvailability.supportedLocales(context)))
            } catch (e: CancellationException) {
                throw e
            } catch (e: AndroidTranscribeError) {
                callback(Result.failure(e.toFlutterError()))
            } catch (e: Exception) {
                callback(
                    Result.failure(
                        AndroidTranscribeError.PlatformError("$e").toFlutterError(),
                    ),
                )
            }
        }
    }

    override fun downloadModel(locale: String) {
        currentDownloadJob?.cancel()
        currentDownloadJob = mainScope.launch { runDownload(locale) }
    }

    override fun transcribeFile(request: TranscribeRequest) {
        // design.md §3のセッション排他(v1では同時1本まで)はDart側
        // `TranscribeSessionGuard`が担うため、通常この時点で前のJobが
        // 生きていることはない。念のため多重起動を防ぐ。
        currentTranscriptionJob?.cancel()
        currentTranscriptionJob = mainScope.launch { runTranscription(request) }
    }

    override fun cancel() {
        currentTranscriptionJob?.cancel()
        currentDownloadJob?.cancel()
    }

    /**
     * `segments` EventChannel自体がキャンセルされた場合の保険
     * (`EventChannelWrappers.kt`のドキュメントコメント参照)。
     *
     * 停止するのは文字起こしJobのみである。2本のEventChannelは独立して
     * いるため、片方の購読解除でもう片方を巻き添えにしてはならない
     * (例: ja-JPの文字起こし中にen-USのダウンロードStreamの購読を解除
     * しても、文字起こしは継続しなければならない)。明示的な[cancel]は
     * 従来どおり両方を停止する。
     */
    fun cancelTranscriptionFromEventChannel() {
        currentTranscriptionJob?.cancel()
    }

    /**
     * `downloadProgress` EventChannel自体がキャンセルされた場合の保険。
     * 停止するのはダウンロードJobのみである(理由は
     * [cancelTranscriptionFromEventChannel]のコメント参照)。
     */
    fun cancelDownloadFromEventChannel() {
        currentDownloadJob?.cancel()
    }

    private suspend fun runDownload(locale: String) {
        // design.md §3細則3: downloadable以外で呼ばれた場合は状態を変化
        // させず何もemitせず完了する。
        val state = ModelAvailability.checkModel(context, locale)
        if (state != ModelState.DOWNLOADABLE) {
            downloadProgressWrapper.sendEndOfStream()
            return
        }

        downloadProgressWrapper.send(fraction = null, completed = false)
        try {
            ModelAcquisition.run(context, locale) { fraction, completed ->
                downloadProgressWrapper.send(fraction = fraction, completed = completed)
            }
            downloadProgressWrapper.sendEndOfStream()
        } catch (e: CancellationException) {
            downloadProgressWrapper.sendError(AndroidTranscribeError.Cancelled)
            downloadProgressWrapper.sendEndOfStream()
        } catch (e: AndroidTranscribeError) {
            downloadProgressWrapper.sendError(e)
            downloadProgressWrapper.sendEndOfStream()
        } catch (e: Exception) {
            downloadProgressWrapper.sendError(AndroidTranscribeError.PlatformError("$e"))
            downloadProgressWrapper.sendEndOfStream()
        }
    }

    private suspend fun runTranscription(request: TranscribeRequest) {
        // design.md §2.2: `playbackRate`はWeb専用オプションであり、
        // Androidでは無視する。PigeonスキーマのTranscribeRequestには
        // 現行ブランチの時点でplaybackRateフィールド自体が存在しない
        // (Darwin実装と同じ事情)ため、path/localeのみを使用する
        // 形で自然に無視される。
        try {
            RecognitionSession.run(context, request, segmentsWrapper)
        } catch (e: CancellationException) {
            // RecognitionSession.run()内のinvokeOnCancellationで既に
            // Cancelledエラーを送出済みのため、ここでは何もしない。
        } catch (e: AndroidTranscribeError) {
            segmentsWrapper.sendError(e)
        } catch (e: Exception) {
            segmentsWrapper.sendError(AndroidTranscribeError.PlatformError("$e"))
        }
    }
}
